ledger/src/ui/src/pin.c

```diff
@@ -108,24 +70,79 @@ unsigned int update_pin_buffer(volatile unsigned int rx, pin_t* pin_ctx) {
 /*
  * Implements RSK NEW PIN command.
  *
- * Sets the device pin.
+ * Sets and checks the device pin.
  *
- * @arg[in] pin_ctx pin context
- * @ret             number of transmited bytes to the host
+ * @ret number of transmited bytes to the host
  */
-unsigned int set_device_pin(pin_t* pin_ctx) {
+unsigned int set_pin() {
 #ifndef DEBUG_BUILD
-    if (!is_pin_valid(pin_ctx)) {
+    if (!is_pin_valid()) {
         THROW(ERR_INVALID_PIN);
     }
 #endif
     // Set PIN
-    os_perso_set_pin(0, GET_PIN(pin_ctx), GET_PIN_LENGTH(pin_ctx));
+    os_perso_set_pin(0, GET_PIN(), GET_PIN_LENGTH());
     // check PIN
     os_global_pin_invalidate();
     unsigned char output_index = CMDPOS;
     SET_APDU_AT(output_index++, 2);
     SET_APDU_AT(output_index++,
-                os_global_pin_check(GET_PIN(pin_ctx), GET_PIN_LENGTH(pin_ctx)));
+                os_global_pin_check(GET_PIN(), GET_PIN_LENGTH()));
     return output_index;
+}
+
+/*
+ * Validates that the pin curently saved to the internal buffer has exactly
+ * PIN_LENGTH alphanumeric characters with at least one alphabetic character.
+ *
+ * @ret     true if pin is valid, false otherwise
+ */
+bool is_pin_valid() {
+    // PIN_LENGTH is the only length accepted
+    if (GET_PIN_LENGTH() != PIN_LENGTH) {
+        return false;
+    }
+    // Check if PIN is alphanumeric
+    bool hasAlpha = false;
+    for (int i = 0; i < PIN_LENGTH; i++) {
+        if (!IS_ALPHANUM(GET_PIN()[i])) {
+            return false;
+        }
+        if (hasAlpha || IS_ALPHA(GET_PIN()[i])) {
+            hasAlpha = true;
+        }
+    }
+
+    return hasAlpha;
+}
+
+/*
+ * Fills the internal pin buffer with zeroes
+ */
+void clear_pin() {
+    explicit_bzero(G_pin_buffer, sizeof(G_pin_buffer));
+}
+
+/*
+ * Validates the pin currently saved to the internal pin buffer
```

Comment:
I would try to be more explicit with the naming and description here so as to not to confuse this with pin rules validation. This method actually unlocks the device, giving access to secrets usage, so I'd try and find a name closer to that description.